stripe-core/src/main/java/com/stripe/android/core/networking/RetryInterceptor.kt

```diff
@@ -0,0 +1,24 @@
+package com.stripe.android.core.networking
+
+import androidx.annotation.RestrictTo
+import java.net.ConnectException
+import java.net.SocketTimeoutException
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+class RetryInterceptor {
+
+    /**
+     * @param request the [StripeRequest] that was made
+     * @param result the [Result] of the request.
+     *  - If the request was successful, this will be the [StripeResponse] that was returned
+     *  - If the request failed, this will be the [Exception] that was thrown
+     * @return `true` if the request can be retried, `false` otherwise
+     */
+    fun <BodyType> shouldRetry(
+        request: StripeRequest,
+        result: Result<StripeResponse<BodyType>>
+    ): Boolean = result.fold(
+        onSuccess = { request.retryResponseCodes.contains(it.code) },
+        onFailure = { it is SocketTimeoutException || it is ConnectException }
```

Comment:
I don't know if it's safe to blanket do this. We certainly wouldn't want to double charge someone if we had a socket timeout, but the server actually processed the request. I think our API design prevents this broadly, but at least to start it seems like a better idea to do this via an opt in, or only for GET requests, etc.

Reply:
looks like we do this on our Java API client (which is not 100% guaranteed to make it safe on mobile SDKs) - https://github.com/stripe/stripe-java/pull/1072/files